### PR TITLE
Refactor improvement checking and adaptive population size in GP Optimisers

### PR DESCRIPTION
**Pull request by gkirgizov**
Primary changes:
- Introduce GenerationKeeper:
  It encapsulates watching improvements/stagnation and keeping best individuals through generations 
- Drop usages of supplementary metrics in optimisers:
  In favor of using compound SingleObjFitness with supplementary metrics.
- Encapsulate dynamic parameter of next offspring size in AdaptiveParameter.
  This simplifies main optimiser loop a lot, allowing to drop many variables with duplicate functionality,
  thanks to combination of GenerationKeeper and Fitness. See `AdaptivePopulationSize` class.
- Other minor code refactorings (notably dropping custom ParetoFront subclass as it doesn't really override any functionality)

Important notes on changes in API:
- Now if only 1 metric is provided, additional default complexity metric is appended to `metrics`
  The appended metric is the previously used and scattered in Optimisers supplementary metric,
  which is used for distinguishing individuals with equal fitness.
  See usages of `ComposerBuilder._default_complexity_metrics()`

This PR is a part of big refactoring #608

### CODE REVIEW

File: fedot/core/optimisers/gp_comp/composite_condition.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+from typing import Optional, List, Callable, Tuple, Iterable
+
+from fedot.core.log import default_log, Log
+
+ConditionType = Callable[[], bool]
+ConditionEntryType = Tuple[ConditionType, Optional[str]]
+
+
+class CompositeCondition:
```
Comment (nicl-nno) [inline fedot/core/optimisers/gp_comp/composite_condition.py:9]: Мне кажется лучше не использовать понятие Composite, чтобы не было путаницы с composite pipelines/composer и т.п

File: fedot/core/optimisers/gp_comp/composite_condition.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+from typing import Optional, List, Callable, Tuple, Iterable
```
Comment (nicl-nno) [inline fedot/core/optimisers/gp_comp/composite_condition.py:1]: Может вынести этот скрипт в другой модуль? Он же ничего специфичного для optimisers.

File: fedot/core/optimisers/gp_comp/generation_keeper.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    :param keep_n_best: How many best individuals to keep from all generations.
+    NB: relevant only for single-objective optimization!
+    """
+
+    def __init__(self,
+                 metrics: Sequence[MetricsEnum] = (),
+                 is_multi_objective: bool = False,
+                 keep_n_best: int = 1,
+                 initial_generation: PopulationT = None):
+        self._generation_num = -1  # -1 means no generations
```
Comment (nicl-nno) [inline fedot/core/optimisers/gp_comp/generation_keeper.py:51]: Может лучше None?

File: fedot/core/optimisers/gp_comp/generation_keeper.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    @property
+    def best_individuals(self) -> Sequence[Individual]:
+        return self.archive.items
+
+    @property
+    def generation_num(self) -> int:
+        return self._generation_num
+
+    @property
+    def stagnation_length(self) -> int:
```
Comment (nicl-nno) [inline fedot/core/optimisers/gp_comp/generation_keeper.py:68]: stagnation_duration мб?

File: fedot/core/optimisers/gp_comp/generation_keeper.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        return self._stagnation_counter
+
+    @property
+    def last_improved(self) -> bool:
```
Comment (nicl-nno) [inline fedot/core/optimisers/gp_comp/generation_keeper.py:72]: Тут нужен или комментарий, или другое название (is_any_improvement?) - непонятно почему last_improved это bool.

File: fedot/core/optimisers/gp_comp/generation_keeper.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        return any(self._metrics_improvement.values())
+
+    def metric_improved(self, metric: MetricsEnum) -> bool:
+        return self._metrics_improvement.get(metric, False)
```
Comment (nicl-nno) [inline fedot/core/optimisers/gp_comp/generation_keeper.py:76]: А это где-то используется?

File: fedot/core/optimisers/gp_comp/generation_keeper.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+    @property
+    def quality_improved(self) -> bool:
```
Comment (nicl-nno) [inline fedot/core/optimisers/gp_comp/generation_keeper.py:79]: Может для единнообразция использовать везде нотацию с is?

File: fedot/core/optimisers/gp_comp/operators/operator.py [hunk 8]
```diff
@@ -25,3 +25,19 @@ class Operator(ABC, Generic[T]):
     @abstractmethod
     def __call__(self, operand: T) -> T:
         pass
+
+
```
Comment (nicl-nno) [inline fedot/core/optimisers/gp_comp/operators/operator.py:29]: Может этот скрипт тогда переименовать в base.py или как-то так? Вроде новый класс не имеет отшноения к операторам.

File: fedot/core/optimisers/gp_comp/param_free_gp_optimiser.py [hunk 9]
```diff
@@ -112,86 +90,72 @@ def optimise(self, objective_function, offspring_rate: float = 0.5,
                                            timer=t)
 
                 if self.parameters.multi_objective:
-                    filtered_archive_items = duplicates_filtration(archive=self.archive,
-                                                                   population=individuals_to_select)
+                    # TODO: feels unneeded, ParetoFront does it anyway
```
Comment (nicl-nno) [inline fedot/core/optimisers/gp_comp/param_free_gp_optimiser.py:93]: Можно попробовать проверить, временно добавив экспешен в случае дубликата.

File: fedot/core/optimisers/gp_comp/param_free_gp_optimiser.py [hunk 10]
```diff
@@ -112,86 +90,72 @@ def optimise(self, objective_function, offspring_rate: float = 0.5,
...
 
                     selected_individuals = selection(types=self.parameters.selection_types,
                                                      population=individuals_to_select,
                                                      pop_size=num_of_parents,
                                                      params=self.graph_generation_params)
 
                     new_population = []
-
                     for ind_1, ind_2 in zip_longest(selected_individuals[::2], selected_individuals[1::2]):
                         new_population += self.reproduce(ind_1, ind_2)
```
Comment (nicl-nno) [inline fedot/core/optimisers/gp_comp/param_free_gp_optimiser.py:110]: Может это вынести в метод базового класса, чтобы не дублировать?

File: fedot/core/optimisers/optimizer.py [hunk 11]
```diff
@@ -102,31 +100,27 @@ def is_equal_fitness(self, first_fitness, second_fitness, atol=1e-10, rtol=1e-10
...
+    # TODO: possibly remove, it's protected & not really used as abstract method
+    @abstractmethod
     def _is_stopping_criteria_triggered(self):
-        is_stopping_needed = self.stopping_after_n_generation is not None
-        if is_stopping_needed and self.num_of_gens_without_improvements == self.stopping_after_n_generation:
-            self.log.info(f'GP_Optimiser: Early stopping criteria was triggered and composing finished')
-            return True
-        else:
-            return False
+        pass
```
Comment (nicl-nno) [inline fedot/core/optimisers/optimizer.py:123]: В целом можно его реализовать - достаточно полезно иметь возможность узнать, срабатывал ли критерий останова.

**Review by nicl-nno — CHANGES_REQUESTED** [review]

File: fedot/core/optimisers/gp_comp/generation_keeper.py (hunk 3 above)
Comment (gkirgizov) [reply fedot/core/optimisers/gp_comp/generation_keeper.py:51]: Нет, это счетчик, он потом просто инкрементится

File: fedot/core/optimisers/gp_comp/generation_keeper.py (hunk 4 above)
Comment (gkirgizov) [reply fedot/core/optimisers/gp_comp/generation_keeper.py:68]: Тоже раздумывал, length не нравилось. Можно и duration

File: fedot/core/optimisers/gp_comp/generation_keeper.py (hunk 6 above)
Comment (gkirgizov) [reply fedot/core/optimisers/gp_comp/generation_keeper.py:76]: Нет, но можно представить случай, где какой-нибудь реализации лагоритма нужно првоерять особую метрику помимо в целом сложности или качества. Логике класса этот метод соответствует, не вижу в нем проблемы

File: fedot/core/optimisers/gp_comp/operators/operator.py (hunk 8 above)
Comment (gkirgizov) [reply fedot/core/optimisers/gp_comp/operators/operator.py:29]: Ты про класс `Operator`? А то я тут с ним ничего в данном PR не менял.
Может, он и абсатрктен, но как раз к оператором и только в ним он и относится. Это в следующих PR, когда сами операторы отрефакторю, станет видно. Пока что это заготовка.

File: fedot/core/optimisers/gp_comp/param_free_gp_optimiser.py (hunk 9 above)
Comment (gkirgizov) [reply fedot/core/optimisers/gp_comp/param_free_gp_optimiser.py:93]: Ага, проверю и отрефакторю, когда доберусь до операторов. Я разделяю PR на отдельные, но TODO оставляю, чтобы не забыть.

File: fedot/core/optimisers/gp_comp/param_free_gp_optimiser.py (hunk 10 above)
Comment (gkirgizov) [reply fedot/core/optimisers/gp_comp/param_free_gp_optimiser.py:110]: Оставлю это на PR когда уже эво операторы буду менять.

File: fedot/core/optimisers/optimizer.py (hunk 11 above)
Comment (gkirgizov) [reply fedot/core/optimisers/optimizer.py:123]: Это нигде не используется, потому как раз и убираю. В грядущем PR я это, кажется, уже совсем дропнул.

Comment (gkirgizov) [reply fedot/core/optimisers/optimizer.py:123]: То есть не очень понятно, зачем это может быть нужно снаружи, а внутри алгоритм это и так знает. Если будет нужно -- то можно будет отдельно сделать, а так я интерфейс Оптимизатора чищу.

File: examples/advanced/multiobj_optimisation.py [hunk 12]
```diff
@@ -10,15 +11,18 @@ def run_classification_multiobj_example(with_plot=True, timeout=None):
...
                                         'with_tuning': False}, seed=42)
     auto_model.fit(features=train_data, target='class')
     prediction = auto_model.predict_proba(features=test_data)
     print(auto_model.get_metrics(target))
     auto_model.plot_prediction()
     if with_plot:
-        auto_model.best_models.show()
+        PipelineEvolutionVisualiser().visualise_pareto(archive=auto_model.best_models,
+                                                       objectives_names=metric_names,
+                                                       show=True)
```
[the remark this answers is not in the document]
Comment (gkirgizov) [reply examples/advanced/multiobj_optimisation.py:25]: в API чьем?

File: fedot/core/optimisers/gp_comp/operators/operator.py (hunk 8 above)
Comment (nicl-nno) [reply fedot/core/optimisers/gp_comp/operators/operator.py:29]: Нет, я по класс AdaptiveParameter(ABC, Generic[T]), который в скрипте operator.py

File: fedot/core/optimisers/gp_comp/generation_keeper.py (hunk 3 above)
Comment (nicl-nno) [reply fedot/core/optimisers/gp_comp/generation_keeper.py:53]: Тогда хотя бы комментарий скорректировать в стиле # -1 means state before initial generation

File: fedot/core/optimisers/gp_comp/generation_keeper.py (hunk 6 above)
Comment (nicl-nno) [reply fedot/core/optimisers/gp_comp/generation_keeper.py]: Ок. 

Но стоит пояснить докстрингами - разницу между metric_improved и _is_metric_kind_improved сходу понять было сложно.

File: fedot/core/optimisers/optimizer.py (hunk 11 above)
Comment (nicl-nno) [reply fedot/core/optimisers/optimizer.py:123]: Сообщение о том что "Early stopping criteria was triggered and composing finished" выводить было бы полезно - чтобы пользователь знал, закончилась ли оптимизация досрочно. Но действительно можно это делать за рамками этого метода.

File: examples/advanced/multiobj_optimisation.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (nicl-nno) [reply examples/advanced/multiobj_optimisation.py:25]: Тут под API имеется в виду верхнеуровневый пользовательский интерфейс https://github.com/nccr-itmo/FEDOT/tree/master/fedot/api.

Т.е. чтобы было какой-то аналог auto_model.best_models.show(), используя которое не требуется вдаваться в детали реализации.

File: fedot/core/optimisers/gp_comp/param_free_gp_optimiser.py (hunk 9 above)
Comment (gkirgizov) [reply fedot/core/optimisers/gp_comp/param_free_gp_optimiser.py:93]: Возник вопрос:
1) а почему эта фильтрация только для multi_objective?
2) фактически, это у нас условие элитизма, то есть мы добавляем всех лучших, причем в любом случае.

Эта ветка кажется ненужной вообще в таком случае. Что думаешь?

File: fedot/core/optimisers/gp_comp/operators/operator.py (hunk 8 above)
Comment (gkirgizov) [reply fedot/core/optimisers/gp_comp/operators/operator.py:29]: Понял. Лучше тогда вынесу этот класс в свой модуль вместе с `popoulation_size.py`

File: fedot/core/optimisers/gp_comp/param_free_gp_optimiser.py (hunk 9 above)
Comment (nicl-nno) [reply fedot/core/optimisers/gp_comp/param_free_gp_optimiser.py:93]: Ок, давай попробуем убрать. В целом действительно, если и удалять дубликаты - то логично это делать по генотипу. 

Мб это какой-то workaround и старая реализация multi_objective падала когда фитнес повторялся.

**Review by nicl-nno — APPROVED** [review]
Только тесты починить.

File: fedot/api/api_utils/params.py [hunk 13]
```diff
@@ -115,7 +115,7 @@ def _parse_input_params(self, input_params: Dict[str, Any]):
             raise ValueError('This type of task is not not supported in API now')
 
         self.task = self.get_task_params(problem, input_params['task_params'])
-        self.metric_name = self.get_default_metric(problem)
+        self.metric_name = self.metric_name or self.get_default_metric(problem)
```
Comment (MorrisNein) [inline fedot/api/api_utils/params.py:118]: Возможно, это уже было в дискуссии, но на какой случай эта заглушка? Если правильно понимаю стек вызова:

Fedot.__init__ -> self.params.initialize_params -> get_initial_params -> _parse_input_params. Т.е. в нынешнем варианте они всегда None на этой строке.

File: fedot/api/main.py [hunk 14]
```diff
@@ -119,7 +121,7 @@ def __init__(self,
         self.target = None
         self.train_data = None
         self.current_pipeline = None
-        self.best_models = None
+        self.best_models: Optional[HallOfFame] = None
```
Comment (MorrisNein) [inline fedot/api/main.py:124]: Здесь типизация выбивается из общей картины. Можем ли задать типизацию и для других параметров?

File: fedot/core/utilities/aggregate_condition.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
+from typing import Optional, List, Callable, Tuple, Iterable
+
+from fedot.core.log import default_log, Log
+
+ConditionType = Callable[[], bool]
+ConditionEntryType = Tuple[ConditionType, Optional[str]]
+
+
+class AggregateCondition:
```
Comment (MorrisNein) [inline fedot/core/utilities/aggregate_condition.py:9]: Это группа из условий поменьше, судя по докстрингу?
Название `AggregateCondition` выглядит неявным. Дословно можно перевести как агрегатное состояние. Может, `ComplexCondition` подойдёт лучше? И ещё в докстринге остался старый вариант `CompositeCondition`.

File: fedot/core/composer/composer_builder.py [hunk 16]
```diff
@@ -76,13 +76,18 @@ def _default_composer_params(self) -> PipelineComposerRequirements:
...
         # Set metrics
         metric_function = ClassificationMetricsEnum.ROCAUC_penalty
-        if self.task.task_type in (TaskTypesEnum.regression, TaskTypesEnum.ts_forecasting):
+        if task.task_type in (TaskTypesEnum.regression, TaskTypesEnum.ts_forecasting):
             metric_function = RegressionMetricsEnum.RMSE
         return [metric_function]
 
+    @staticmethod
+    def _default_complexity_metrics() -> List[MetricsEnum]:
+        return [ComplexityMetricsEnum.node_num]
```
Comment (Dreamlone) [inline fedot/core/composer/composer_builder.py:89]: Не до конца уверен, но просто выскажу свою мысль. Насколько мне известно, где-то в недрах оптимизации есть механизм регуляризации, который отвечает примерно за то же самое, что и данная метрика - он не позволяет пайплайну разрастаться. Не приведёт ли такое "двойное воздействие" со сторону дополнительной метрики, штрафующей за сложность и нативной регуляризации к тому, что FEDOT перестанет выращивать сколько-нибудь крупные пайплайны?

File: test/unit/optimizer/test_generation_keeper.py [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+
+
+def test_archive_no_improvement():
+    archive = generation_keeper(population1())
+    assert archive.stagnation_duration == 0
+    assert archive.is_any_improved
+    assert archive.is_quality_improved and archive.is_complexity_improved
+    assert archive.generation_num == 0
+
+    archive.append(population1())
```
Comment (MorrisNein) [inline test/unit/optimizer/test_generation_keeper.py:54]: Тут снова по именованиям, в Федоте принято использовать имена переменных по типу `population_1`, `population_2`. Да и читается лучше в целом. Меня уже поправляли по этому поводу, передаю эстафету)

**Review by MorrisNein — CHANGES_REQUESTED** [review]

File: fedot/core/utilities/aggregate_condition.py (hunk 15 above)
Comment (MorrisNein) [reply fedot/core/utilities/aggregate_condition.py:9]: Ещё вариант `ConditionsGroup`, как наиболее явный

**Review by maypink — APPROVED** [review]

**Review by Dreamlone — APPROVED** [review]

File: fedot/core/composer/composer_builder.py (hunk 16 above)
Comment (gkirgizov) [reply fedot/core/composer/composer_builder.py:89]: Не должно привести к этому.
1. Эта метрика не столько штрафует, сколько позволяет выбирать более простую модель среди одинаково эффективных (по первой метрике) моделей. То есть эта метирка используется в `SingleObjFitness` просто как `supplementary metric`. 
2. Эта логика не нова -- эта вторая метрика как бы "вручную" добавлялась внутри эволюционных оптимизаторов. Тут я просто вынес ее на уровень композера и сделал ее явной.

Comment (Dreamlone) [reply fedot/core/composer/composer_builder.py:89]: Принял, вопрос снят

File: test/unit/optimizer/test_generation_keeper.py (hunk 17 above)
Comment (gkirgizov) [reply test/unit/optimizer/test_generation_keeper.py:54]: эстафету принял

File: fedot/api/api_utils/params.py (hunk 13 above)
Comment (gkirgizov) [reply fedot/api/api_utils/params.py:118]: Ты прав, а я уже забыл, чем руководствовался, когда добавил это. Откачу эту строчку.
